packages/cli/src/rollup-plugins/typescript-declarations/common.ts

```diff
@@ -195,11 +195,32 @@ export const getDeclarationsForFile = async (
         (context) => (
           node
         ): import("typescript").Bundle | import("typescript").SourceFile => {
-          if (!visitModuleSpecifier) {
-            return node;
-          }
-
-          const cachedVisitModuleSpecifier = memoize(visitModuleSpecifier);
+          const replaceTsExtensions = (moduleSpecifier: string) => {
+            if (!moduleSpecifier.startsWith(".")) {
+              return moduleSpecifier;
+            }
+            return moduleSpecifier.replace(
+              /(\.d)?\.(ts|tsx|mts|cts)$/,
+              (match) => {
+                return {
+                  ".ts": ".js",
+                  ".tsx": ".js",
+                  ".mts": ".mjs",
+                  ".cts": ".cjs",
+                  ".d.ts": ".js",
+                  ".d.tsx": ".js", // `.d.tsx` is not quite a thing but the current regex matches it, so it's included here for completeness
+                  ".d.mts": ".mjs",
+                  ".d.cts": ".cjs",
+                }[match]!;
+              }
+            );
+          };
+          const cachedVisitModuleSpecifier = memoize(
+            visitModuleSpecifier
+              ? (moduleSpecifier) =>
+                  replaceTsExtensions(visitModuleSpecifier(moduleSpecifier))
+              : replaceTsExtensions
+          );
```

Comment:
Since `allowImportingTsExtensions: true` is allowed with just any `moduleResolution` type ( https://github.com/microsoft/TypeScript/pull/52230 ), I think it's the best/easiest to just apply this logic at all times. 

We could cross-reference if that option is enabled but it seems like an overcomplication to me. If it's not enabled then typechecking process would complain about it and it doesn't quite matter if we replace those or not when generating declaration files